### DISCUSSION TITLE
How to deal with types exposed by assemblies without proper nullable support?

### DISCUSSION THREAD

**Discussion by wizofaus** [dotnet/csharplang#7828] [General]
The type PnP.Core.Model.IExpandoDataModel in PnP.Core includes an indexing operator that is seen as `object this[string key] { get; set; }` by my project (that has nullable enabled). But it seems that this assembly doesn't properly honour the "nullable" directive, and actually does return nulls here.
Should there be a way to reference an assembly in such a way that means all reference types from it are assumed to actually be nullable? (i.e. it should be seen from my project as `object? this[string? key]` - in this case it arguably would be nicer to have it as "string key", but there's no way to automatically determine that "null" isn't a valid value for "key" here).
This problem isn't hypothetical BTW - my app crashed with a NullReferenceException because of this, despite having all nullable-related warnings as errors, and not using any mechanisms to bypass nullable checking etc.

**Comment by huoyaoyuan** [comment]
> it should be seen from my project as `object? this[string? key]`

In that case, you should see the strictest assumption - outputs are nullable and inputs are not nullable. `ref` parameters should be not nullable for input, and may become nullable after every invocation.

It's also harder to cover polymorphism and overrides.

**Reply by wizofaus** [reply]
In this case I don't care if the input parameter is considered nullable or not, but my question is to how to tell the compiler that key reference types used in any interfaces exposed by the assembly should be considered actually nullable (i.e. ```object?```).
